Generating an mdBook-based variant of the SRS - UPDATED

BilalM04 opened the pull request:
This PR is an extension of #3814. Please read https://github.com/JacquesCarette/Drasil/pull/3814#issue-2365220779 before reading this PR.

Contributes to #3722 
Closes #3827 

## Updates / New Additions

- Refactored how the sections are separated (more info below).
- Generated a `SUMMARY.md` file, which is used for the side panel in mdBook.
- Generated a `book.toml` config file, similar to how the CSS is currently generated.
- Generated a title page file (`title.md`)
- Separated into two new SRS formats: Markdown and mdBook (more info below).
- Got the cross-file references working for the mdBook format.
- Fixed minor issues in expression printing.
- Generated SRS for all examples and updated stable. I roughly scanned over and ran all of them. They seem to all be working and rendering as desired.
- Added mdBook installation instructions to the New Workspace Setup wiki.
- Added makefile targets to build and run server of mdBook projects.
  - To build all mdBook projects run `make mdbook_build`. 
  - To build a specific example, run `make X_mdbook_build`. 
  - To launch the local server for a specific example, run `make X_mdbook_server`.

## Refactoring the Separation of Sections

Please read https://github.com/JacquesCarette/Drasil/discussions/3796#discussion-6809343 to understand how SRS generation currently works in Drasil. To summarize, here is the main info:

> The format specific gen functions receive a single Document, which contains all the SRS details. The Document it receives is in the format of the data type defined in [Language.Drasil.Document](https://github.com/JacquesCarette/Drasil/blob/bfcf92f09b07bd883ef2af438f300499d35305e0/code/drasil-lang/lib/Language/Drasil/Document.hs#L61-L64), where the SRS contents is represented through [Section]. The format specific gen functions, using the [makeDocument](https://github.com/JacquesCarette/Drasil/blob/bfcf92f09b07bd883ef2af438f300499d35305e0/code/drasil-printers/lib/Language/Drasil/Printing/Import/Document.hs#L22-L27) function, then convert it to the Document data type defined in [Language.Drasil.Printing.LayoutObj](https://github.com/JacquesCarette/Drasil/blob/bfcf92f09b07bd883ef2af438f300499d35305e0/code/drasil-printers/lib/Language/Drasil/Printing/LayoutObj.hs#L9-L10), where the SRS content is represented through [LayoutObj]. The SRS printers then render these LayoutObj's into a single Doc.

In #3814, the SRS was separated into separate files once it became in the format of the `Document` data type defined in [`Language.Drasil.Printing.LayoutObj`](https://github.com/JacquesCarette/Drasil/blob/bfcf92f09b07bd883ef2af438f300499d35305e0/code/drasil-printers/lib/Language/Drasil/Printing/LayoutObj.hs#L9-L10). I found this to be quite a late stage to be separating, as it was hard to distinguish where each `LayoutObj` belonged. Additionally, the `Document` type is implicitly assuming a single file is being generated. Instead I created a new data type `Project` and `File`:

https://github.com/JacquesCarette/Drasil/blob/5f9ba9280a05c8eca11af14c06ef3957e1fdc64f/code/drasil-printers/lib/Language/Drasil/Printing/LayoutObj.hs#L14-L16

The `Project` type contains a `RefMap`, which maps reference anchor ids to the `Filename` of the `File` that contains them. This map is then used when rendering references in mdBook.

Unlike in #3814, where the given SRS gets converted to a `Document` type, it now gets [converted to a `Project`](https://github.com/JacquesCarette/Drasil/blob/5f9ba9280a05c8eca11af14c06ef3957e1fdc64f/code/drasil-printers/lib/Language/Drasil/Markdown/Print.hs#L61-L63) using the new [`makeProject`](https://github.com/JacquesCarette/Drasil/blob/5f9ba9280a05c8eca11af14c06ef3957e1fdc64f/code/drasil-printers/lib/Language/Drasil/Printing/Import/Document.hs#L31-L38) function.

## Markdown vs mdBook

I created two new SRS formats. The first being `MDBook` which generates the mdBook project. The other being `Markdown` which generates a single-page SRS in Markdown which can be rendered directly in the GitHub repo. I spoke to @balacij about this, and he seems to agree that having the two separate formats is beneficial. @smiths and @JacquesCarette, what are your thoughts?

The `Markdown` format is not complete. Currently it renders with the exact same syntax as mdBook; however, there are slight differences in syntax between the two. I will not be addressing that issue in this PR/branch as I believe creating a separate issue and branch after this is merged is the best approach. Addressing this `Markdown` format will also help with the issues in the `JSON` format, as the two formats use the same "flavour" of Markdown. 

## Next Steps

- Work with @balacij to get the required assets into the generated `src` directory and host the mdBook SRSs.
- Address the issues in the `Markdown` format.

samm82 commented on code/drasil-gen/lib/Language/Drasil/Generate.hs:
```diff
@@ -7,7 +7,8 @@ module Language.Drasil.Generate (
   -- * Types (Printing Options)
-  DocType(..), DocSpec(DocSpec), Format(TeX, HTML, JSON), DocChoices(DC),
+  DocType(..), DocSpec(DocSpec), DocChoices(DC),
+  Format(TeX, HTML, JSON, Markdown, MDBook),
```
This may be a bit late, but is there a reason we went with `Markdown` instead of `MD` to be consistent with the use of abbreviations for the other formats? I realize that could lead to confusion with `MD` vs. `MDBook`, which might be enough of a reason.

BilalM04 replied:
There was no particular reason, I just went with the first thing that popped up in my head. I agree, there could be confusion between `MD` and `MDBook`. I am open to suggestions :)

BilalM04 replied:
`Jupyter` is actually used as a `DocType` (I presume it is for the lessons). `JSON` is used for the printer format.

https://github.com/JacquesCarette/Drasil/blob/6a86cda80c866f216d7cd663d4876850f4e971fe/code/drasil-printers/lib/Language/Drasil/Format.hs#L4-L10

BilalM04 replied:
Yes, using `Jupyter` for both is a problem. We could use `Jupyter` for `DocType` and `Ipynb` for `Format`.

```
error:
drasil-printers>     Multiple declarations of ‘Jupyter’
drasil-printers>     Declared at: lib/Language/Drasil/Format.hs:7:32
drasil-printers>                  lib/Language/Drasil/Format.hs:10:36
drasil-printers>    |
drasil-printers> 10 | data Format = TeX | Plain | HTML | Jupyter | Markdown | MDBook
drasil-printers>    |                                    ^^^^^^^
```

samm82 replied:
I personally like that!

BilalM04 replied:
This was discussed in #3815. 

- `DocType`: `Jupyter` will be changed to `Lesson`
- `Format`: `JSON` will be changed to `Jupyter`
- `Markdown` will be extracted into a new data type, representing the different Markdown flavours.

BilalM04 replied:
Added in https://github.com/JacquesCarette/Drasil/pull/3823/commits/f987ff95bb87561978b387243c83b98742758ed1

samm82 commented on code/drasil-gen/lib/Language/Drasil/Generate.hs:
```diff
@@ -70,7 +85,8 @@ prntDoc' dt dt' fn format body' sm = do
         getExt JSON = ".ipynb"
-        getExt _    = error "We can only write in TeX, HTML and Jupyter Notebook (for now)."
+        getExt Markdown = ".md"
+        getExt _    = error "We can only write in TeX, HTML, Jupyter Notebook, Markdown, and mdBook (for now)."
```
A comment here saying that the `MDBook` case is handled above is probably a good idea, for those who don't see it in the previous pattern matching and think there's a mistake (couldn't be me…)

BilalM04 replied:
Fixed in https://github.com/JacquesCarette/Drasil/pull/3823/commits/f7ba8fda90132e900e5349385606160f9c09ac72.

samm82 commented on code/drasil-gen/lib/Language/Drasil/Generate.hs:
```diff
@@ -89,12 +105,27 @@ prntCSS docType fn body = do
--- | Renders the documents.
+-- | Helper for generating the .toml config file for mdBook.
+printBook :: DocType -> Document -> PrintingInformation -> IO()
+printBook dt doc sm = do
```
Why `printBook` instead of `prntBook` (without an "i") for consistency with other functions? Or, better yet, why is `prnt` used in other functions?

BilalM04 replied:
Fixed in https://github.com/JacquesCarette/Drasil/pull/3823/commits/f7ba8fda90132e900e5349385606160f9c09ac72.

samm82 commented on code/drasil-gen/lib/Language/Drasil/Generate.hs:
```diff
@@ -89,12 +105,27 @@ prntCSS docType fn body = do
+-- | Renders multi-page documents.
+writeDoc' :: PrintingInformation -> Format -> Document -> [(String, Doc)]
+writeDoc' s MDBook doc = genMDBook s doc
+writeDoc' _ _      _   = error "we can only write TeX/HTML/JSON/Markdown/MDBook (for now)"
```
This error string should be made consistent and reused instead of manually defining it repeatedly

BilalM04 replied:
Fixed in https://github.com/JacquesCarette/Drasil/pull/3823/commits/f7ba8fda90132e900e5349385606160f9c09ac72.

samm82 commented on code/drasil-printers/lib/Language/Drasil/Markdown/Helpers.hs:
```diff
@@ -0,0 +1,111 @@
+
+-- | Helper for setting up links to external URIs
+reflinkURI :: Doc -> Doc -> Doc
+reflinkURI ref txt = if ref==txt then ang ref
```
```suggestion
reflinkURI ref txt = if ref == txt then ang ref
```

I _feel_ like this is better formatting, unless I'm missing something. If I'm not, I'm surprised HLint didn't complain!

samm82 commented on code/drasil-printers/lib/Language/Drasil/Markdown/Helpers.hs:
```diff
@@ -0,0 +1,111 @@
+
+h :: Int -> Doc
+h n
+  | n < 1     = error "Illegal header (too small)"
+  | n > 4     = error "Illegal header (too large)"
```
Why are these too small/large? More context would be helpful to the user

BilalM04 replied:
Fixed in https://github.com/JacquesCarette/Drasil/pull/3823/commits/6a86cda80c866f216d7cd663d4876850f4e971fe.

samm82 commented on code/drasil-printers/lib/Language/Drasil/Markdown/Helpers.hs:
```diff
@@ -0,0 +1,111 @@
+h n
+  | n < 1     = error "Illegal header (too small)"
+  | n > 4     = error "Illegal header (too large)"
+  | n < 4     = text "#"
```
Is this correct? I would have expected instances of `"##"` or and `"###"`, probably created by something like `"#" * n`

BilalM04 replied:
In the SRS declaration, the heading weights are defined assuming a single file will be rendered. However, mdBook extracts the subsections and subsubsections into their own separate sections. Thus, if I defined the headings to be `"#" * n`. the subsections and subsubsections (which are now separate sections) will have smaller headings, and the heading sizes will be inconsistent across the various `.md` files.

This will ultimately be addressed when fixing up the rendering of the single-page Markdown SRS, as that will require the headings to be rendered in different sizes, consistent with the SRS declaration.

samm82 reviewed (changes requested):
Didn't have time to look through everything, but here's my progress on my classic nitpicking. 😂 Super minor notes on a really solid PR!

balacij commented on code/drasil-printers/lib/Language/Drasil/Markdown/Book.hs:
```diff
@@ -0,0 +1,30 @@
+
+-- | Generates the .toml config file for mdBook.
+makeBook :: Document -> PrintingInformation -> Doc  
+makeBook (Document t _ _ _) sm = vcat [
```
Is it worth creating a record data type that carries the [configuration (according to their spec)](https://rust-lang.github.io/mdBook/format/configuration/general.html) before building the TOML?

BilalM04 replied:
As of now, all the TOMLs will be identical, apart from the title. I kept the TOML quite simple, only defining the absolute minimum needed for the book to run as desired. I don't anticipate a need to use the other configuration options.

balacij commented on code/drasil-printers/lib/Language/Drasil/Markdown/Helpers.hs:
```diff
@@ -0,0 +1,112 @@
+wrap a = wrapGen hcat Id a empty
+
+-- | Helper for setting up HTML tags
+wrapGen :: ([Doc] -> Doc) -> Variation -> String -> Doc -> [String] -> Doc -> Doc
```
Can you please add some more documentation to this?

BilalM04 replied:
Added in https://github.com/JacquesCarette/Drasil/pull/3823/commits/3db0792f58ac67a73664cbb1da3e768558f3f447. Additionally, I modified the type signature to remove an unnecessary parameter.

balacij commented on code/drasil-printers/lib/Language/Drasil/Markdown/Helpers.hs:
```diff
@@ -0,0 +1,112 @@
+
+-- | Curly braces.
+br :: Doc -> Doc
+br x = text "{" <> x <> text "}"
```
Are these functions duplicates of any in `Language.Drasil.Printing.Helpers`?

BilalM04 replied:
The functions in `Language.Drasil.Printing.Helpers` are `String -> Doc`

https://github.com/JacquesCarette/Drasil/blob/6a86cda80c866f216d7cd663d4876850f4e971fe/code/drasil-printers/lib/Language/Drasil/Printing/Helpers.hs#L48-L53

I can either:
1) Keep it the way it is.
2) Remove the functions in `Language.Drasil.Markdown.Helpers` and use `show` on the parameter being passed in.

Which do you suggest?

balacij commented on code/drasil-printers/lib/Language/Drasil/Markdown/Helpers.hs:
```diff
@@ -0,0 +1,112 @@
+
+-- | Helper for stripping Docs
+stripStr :: Doc -> Doc -> Doc
+stripStr d s = hcat (map text (splitOn (show s) (show d)))
```
This seems a bit odd -- it converts from a `Doc` to a String and then re-renders it back into a `Doc`. Is there any way we can avoid this?

BilalM04 replied:
The function is used to strip `Doc`s of any `\n` and/or `\t`. `Doc`s are stripped of `\n` when rendering the Markdown tables as the `printLO` function prints to separate lines; however, `LayoutObj`s in the tables need to be on a single line in order to be rendered. `Doc`s are stripped of `\t` as when calling `TeX.pExpr`, the TeX expressions return formatted expressions, which look ugly in Markdown.

To answer your question, no. Since the printers all deal with `Doc`s, and there are no existing functions in `Text.PrettyPrint` to strip `Doc`s of certain characters, the conversion is required.

balacij commented on code/drasil-printers/lib/Language/Drasil/Markdown/Print.hs:
```diff
@@ -0,0 +1,344 @@
+  where
+    lo = lookup "Label" ps
+    c = maybe l (\lo' -> makeLO rm ("Label", lo')) lo 
+    header = nl <> text "##" <+> heading c l <> nl
```
Instead of using `text "##"`, I think it wold be good to use `h2` or something of the sorts.

BilalM04 replied:
Addressed in https://github.com/JacquesCarette/Drasil/pull/3823/commits/88a109896dbccd86b0a148837adea09cb8b556a7. I created a `h'` function which fixes this issue and also creates a useful function that can be used when fixing the `Markdown` format rendering (also mentioned in https://github.com/JacquesCarette/Drasil/pull/3823#discussion_r1657652245)

balacij commented on code/stable/dblpend/SRS/Markdown/DblPend_SRS.md:
```diff
@@ -0,0 +1,1128 @@
+|Source     |[accelerationWiki](#accelerationWiki)                                                                                                                                                                                                                           |
+|RefBy      |                                                                                                                                                                                                                                                                |
+
+<div align="center">
+
+## Velocity {#TM:velocity}
+
+</div>
```
Does this work in most Markdown renderers?

BilalM04 replied:
The heading attributes (`{#id}`) are specific to mdBook, other Markdown renderers require an `id` attribute in the tag. Additionally, the definition headings are aligned to the center as tables in mdBook are centered. For other Markdown renderers aligning the heading to the center is not needed, as all elements are left-justified.

balacij replied:
Since we know that this is the GHFM flavour, this works. So, this comment can be dismissed. Thank you @BilalM04

balacij commented on code/stable/dblpend/SRS/Markdown/DblPend_SRS.md:
```diff
@@ -0,0 +1,1128 @@
+
+<div align="center">
+
+## Velocity {#TM:velocity}
```
I think that custom IDs are only supported in extended Markdown flavours, so it might be worth noting around the definition of the printer that it generates an extended Markdown with MathJax and custom header links.

BilalM04 replied:
Note added in https://github.com/JacquesCarette/Drasil/pull/3823/commits/5a025acd47ca6c3df8bf89c42c6dbd1aa7c84a00.

balacij commented on code/stable/dblpend/SRS/mdBook/src/SecSysContext.md:
```diff
@@ -0,0 +1,21 @@
+
+<div id="Figure:sysCtxDiag"></div>
+
+![System Context](../../../../datafiles/dblpend/SystemContextFigure.png)
```
Did images end up working if they weren't within the `src` folder of the mdBook project?

BilalM04 replied:
Nope, after researching online it seems the only way to render images is by having a copy in the `src` folder. After solving that issue, the generated links will also have to be updated to the new paths.

BilalM04 commented on code/drasil-printers/lib/Language/Drasil/Markdown/Print.hs:
```diff
@@ -0,0 +1,344 @@
+-------------------- EXPRESSION PRINTING ------------------------
+-----------------------------------------------------------------
+
+-- | Helper for rendering Exprs into mdBook compatiable Mathjax
+pExpr :: Expr -> Doc
+pExpr (Str s)        = printMath $ toText $ pure $ lq <> text s <> rq
+  where 
+    lq = text "\\\\(\\``\\\\)"
+    rq = text "''"
+pExpr (Div n d)      = printMath $ command2D "frac" (pExpr' n) (pExpr' d)
+pExpr (Case ps)      = printMath $ mkEnv "cases" (TeX.cases ps lnl pExpr')
+pExpr (Mtx a)        = stripStr (printMath $ mkEnv "bmatrix" (TeX.pMatrix a lnl pExpr')) tab
+pExpr (Row [x])      = br $ pExpr x 
+pExpr (Row l)        = foldl1 (<>) (map pExpr l)
+pExpr (Label s)      = printMath $ TeX.pExpr (Label s')
+  where s' = replace "*" "\\*" (replace "_" "\\_" s)
+pExpr (Sub e)        = bslash <> unders <> br (pExpr e)
+pExpr (Sup e)        = hat    <> br (pExpr e)
+pExpr (Over Hat s)   = printMath $ commandD "hat" (pExpr' s)
+pExpr (MO Perc)      = bslash <> printMath (TeX.pExpr (MO Perc))
+pExpr (Fenced l r m) = fence TeX.Open l <> pExpr m <> fence TeX.Close r
+  where 
+    fence _ Abs = text "\\|"
+    fence a b   = printMath $ TeX.fence a b
+pExpr (Font Bold e)  = printMath $ commandD "boldsymbol" (pExpr' e)
+pExpr (Font Emph e)  = pExpr e
+pExpr (Spc Thin)     = text "\\\\,"
+pExpr (Sqrt e)       = printMath $ commandD "sqrt" (pExpr' e)
+pExpr e              = printMath $ TeX.pExpr e
```
I had a discussion with @balacij about this, I will summarize it here.

At first glance, the code for `pExpr` seems very similar to the one in [`Language.Drasil.TeX.Print`](https://github.com/JacquesCarette/Drasil/blob/88a109896dbccd86b0a148837adea09cb8b556a7/code/drasil-printers/lib/Language/Drasil/TeX/Print.hs#L101-L124) and one may assume that it is duplicated. However, there is a reason for this. The differences in expressions between pure LaTeX and the version of Mathjax that mdBook uses is very minor. Specifically, the differences are extra escape characters, different command names (i.e. boldsymbol), quotes, etc. I took the approach of re-defining all the expressions that are different, then delegating to the TeX function in a catch all statement. But since the `pExpr` functions involve recursive calls I had to re-define those as well. This is because once `TeX.pExpr` is called, there is no way for it to return to the Markdown specific `pExpr`.

The reason I went with this approach rather than another (i.e. adding a new parameter to `TeX.pExpr`) is because it is simple and the least amount of code. Adding an additional parameter to `pExpr` that takes in the recursive function call would result in more functions to be added, such as one to begin the recursion. Additionally since all of the printers leverage `TeX.pExpr`, this change would propagate to all the printers.

I will annotate the code below with my reasoning to why those patterns are re-defined:

```Haskell
-- | Helper for rendering Exprs into mdBook compatiable Mathjax
pExpr :: Expr -> Doc
pExpr (Str s)        = printMath $ toText $ pure $ lq <> text s <> rq -- | quotes are different
  where 
    lq = text "\\\\(\\``\\\\)"
    rq = text "''"
pExpr (Div n d)      = printMath $ command2D "frac" (pExpr' n) (pExpr' d) -- | recursive call
pExpr (Case ps)      = printMath $ mkEnv "cases" (TeX.cases ps lnl pExpr') -- | different parameters to TeX.cases 
pExpr (Mtx a)        = stripStr (printMath $ mkEnv "bmatrix" (TeX.pMatrix a lnl pExpr')) tab -- | different parameters to TeX.pMatrix
pExpr (Row [x])      = br $ pExpr x -- | recursive call
pExpr (Row l)        = foldl1 (<>) (map pExpr l) -- | recursive call
pExpr (Label s)      = printMath $ TeX.pExpr (Label s')  -- | escaping special chars in string
  where s' = replace "*" "\\*" (replace "_" "\\_" s)
pExpr (Sub e)        = bslash <> unders <> br (pExpr e) -- | extra escape char
pExpr (Sup e)        = hat    <> br (pExpr e) -- | recursive call
pExpr (Over Hat s)   = printMath $ commandD "hat" (pExpr' s) -- | recursive call
pExpr (MO Perc)      = bslash <> printMath (TeX.pExpr (MO Perc)) -- | extra escape char
pExpr (Fenced l r m) = fence TeX.Open l <> pExpr m <> fence TeX.Close r -- | different fence and recursive call
  where 
    fence _ Abs = text "\\|"
    fence a b   = printMath $ TeX.fence a b
pExpr (Font Bold e)  = printMath $ commandD "boldsymbol" (pExpr' e) -- | change to 'boldsymbol'
pExpr (Font Emph e)  = pExpr e -- | recursive call
pExpr (Spc Thin)     = text "\\\\," -- | extra escape chars
pExpr (Sqrt e)       = printMath $ commandD "sqrt" (pExpr' e) -- | recursive call
pExpr e              = printMath $ TeX.pExpr e -- | catch all statement
```

The annotation above is not in the actual code, please let me know if it is beneficial to add it to the source code as well, along with key points from my blurb above.

samm82 commented:
When trying to `checkout` this branch, the wiki bug from #3793 seems to reappear - any ideas why? @BilalM04 @balacij 

```bash
samcr@Apex MINGW64 ~/Desktop/Drasil/code (main)
$ git checkout mdBookPrinterTest 
error: invalid path 'wiki/Attributes-and-Chunks:-Meeting-June-7th,-2023.md'
```

samm82 commented:
> To run the mdBook projects, go to the `build/example_name/SRS/mdBook` directory and enter `mdbook serve --open` in terminal to run a local server and view the SRS.

Could we also generate a Makefile to handle this? Idk how involved that would be 😅

BilalM04 commented:
@samm82 the invalid path error is likely due to the fix not being on this branch. I just merged `main` into the `mdBookPrinterTest` branch, which should hopefully resolve the issue.

samm82 commented:
I needed to `git fetch` in order to actually get the new changes locally, as opposed to the `git checkout <branch>; git pull` workflow I'm used to (just thought I'd share my silliness). I'm able to switch over to the branch now! 🙃

BilalM04 commented:
From @samm82's suggestions in https://github.com/JacquesCarette/Drasil/pull/3823#issuecomment-2197367526, I added the mdBook installation instructions to the New Workspace Setup wiki in https://github.com/JacquesCarette/Drasil/pull/3823/commits/b9b2047435320e126c2812f58fa152e1042736af, and added makefile targets for mdBook in https://github.com/JacquesCarette/Drasil/pull/3823/commits/df11d396354854a16b70cb3970e65d65c85f8ebf. @balacij suggested to keep these additions within this PR.

Added makefile targets:

- `make mdBook_build` - Build all generated mdBook projects.
- `make X_mdBook_build` - Build the mdBook SRS of an individual example.
- `make X_mdBook_server` - Launch local server for the mdBook SRS of an individual example.

samm82 commented:
Should these be renamed to `mdbook` (all lowercase)? That's what the command is, which is what I assumed the `make` rule would be called (if there's a reason for using `mdBook`, I'm sure that's fine too!

BilalM04 commented:
I used `mdBook` as that is how it is capitalized in the [mdBook documentation](https://rust-lang.github.io/mdBook/index.html). However, I agree it should be renamed to `mdbook` in the makefile, as the actual command is all lower case. Additionally it will now be consistent with all the other targets (which are also all lower case).

samm82 commented on code/drasil-printers/lib/Language/Drasil/Markdown/Helpers.hs:
```diff
@@ -0,0 +1,127 @@
+  | n == 2    = text "##"
+  | n == 3    = text "###"
+  | n == 4    = text "####"
+  | otherwise = error "Illegal header (header weight must 0 < n < 5)"
```
```suggestion
  | otherwise = error "Illegal header (header weight must be between 0 and 5)"
```
Just to be consistent with the other error messages which don't use `n`, although this could be approached in other ways

BilalM04 replied:
Fixed in https://github.com/JacquesCarette/Drasil/pull/3823/commits/20ebf35520a935854188b3cab2b4600062620bc3.

BilalM04 commented on code/drasil-printers/lib/Language/Drasil/Markdown/Helpers.hs:
```diff
@@ -0,0 +1,127 @@
+-- | Helper for setting up heading weights in mdBook.
+h :: Int -> Doc
+h n
+  | n < 1     = error "Illegal header (header weigth must be > 0)."
```
Fixed in https://github.com/JacquesCarette/Drasil/pull/3823/commits/20ebf35520a935854188b3cab2b4600062620bc3.

samm82 reviewed (dismissed):
This seems good to me (once @balacij's items have been addressed) 😁

samm82 commented:
@BilalM04 I'd make a separate issue for each of the notes for future work @JacquesCarette made so that we don't lose track of them/can have more detailed (and contained!) discussion on them!

BilalM04 commented:
In regards to the `splitStr` function and `\n` and `\t` appearing in `Doc`s. This was due to the fact `TeX.pExpr` returns case statements and matrices formatted to separate lines. However in mdBook, they need to be in a single line to be rendered (particularly for the tables). Thus, I was naively converting to a `String` and removing all the formatting. In https://github.com/JacquesCarette/Drasil/pull/3823/commits/feec894a051213d89167756eef8a486fa4bf200c, from @balacij's suggestion, I parameterized the concatenation functions in `mkEnv`, `TeX.pMatrix`, and `TeX.cases`, and passed in horizontal concatenation in the Markdown printers.

Additionally, in order to add empty lines in the `Doc`s, I was using `text "\n"`, which I realized to be the wrong approach. In https://github.com/JacquesCarette/Drasil/pull/3823/commits/dc5954be09c49b4fefb7143bf0a9fba109ee472b, I added a new infix operator which concatenates two `Doc`s with an empty line in between them.

balacij commented on code/Makefile:
```diff
@@ -371,6 +379,16 @@ $(filter %$(STABILIZE_E_SUFFIX), $(STABILIZE_WEBSITE)): %$(STABILIZE_E_SUFFIX):
 
+# Build individual example mdBook SRSs.
+# Targets will have the form exampleName_mdBook_build.
+$(filter %$(MDBOOK_BUILD_E_SUFFIX), $(MDBOOK_BUILD_EXAMPLES)): %$(MDBOOK_BUILD_E_SUFFIX): $(CLEAN_GF_PREFIX)$(LOG_FOLDER_NAME) %$(GEN_E_SUFFIX)
+	- mdbook build "$(BUILD_FOLDER)$(EDIR)/SRS/mdBook"
+
+# Launch local server for individual example mdBook SRSs.
+# Targets will have the form exampleName_mdBook_server.
+$(filter %$(MDBOOK_SERVER_E_SUFFIX), $(MDBOOK_SERVER_EXAMPLES)): %$(MDBOOK_SERVER_E_SUFFIX): $(CLEAN_GF_PREFIX)$(LOG_FOLDER_NAME) %$(GEN_E_SUFFIX)
+	- mdbook serve --open "$(BUILD_FOLDER)$(EDIR)/SRS/mdBook"
```
The `-` in front suppresses error codes. I think we want those to be returned (I previously added too many of the `-` in the Makefile, I need to fix those too). So, we should remove the prepended `-`s.

BilalM04 replied:
Fixed in https://github.com/JacquesCarette/Drasil/pull/3823/commits/6eaf3dccd8fd7da4b86aa5df17a5bfd496897c68.

balacij commented on code/drasil-gen/lib/Language/Drasil/Generate.hs:
```diff
@@ -52,25 +55,46 @@ prntDoc :: Document -> PrintingInformation -> String -> DocType -> Format -> IO
+    TeX               -> do prntDoc' dtype (show dtype ++ "/PDF") fn TeX d pinfo
+                            prntMake $ DocSpec (DC dtype []) fn
+    Jupyter           -> do prntDoc' dtype (show dtype ++ "/Jupyter") fn Jupyter d pinfo
+    (Markdown GitHub) -> do prntDoc' dtype (show dtype ++ "/Markdown") fn (Markdown GitHub) d pinfo
```
`Markdown GitHub` to show that it uses the GHFM spec => Good idea!

balacij commented on code/drasil-printers/package.yaml:
```diff
@@ -14,6 +14,7 @@ dependencies:
 - lens
 - pretty
 - containers
+- MissingH
```
Out of curiosity, where is this used?

BilalM04 replied:
It is used [here](https://github.com/JacquesCarette/Drasil/pull/3823/files?diff=unified&w=0#diff-1247766c67d1c8c19d75aacc0f2153498b823b7055a0aac385c064ae5f66cdb5R155). For the `replace` function imported from `Data.List.Utils`.

balacij replied:
Ah, gotcha!

balacij reviewed (approved):
Great work! The PR looks like it's in great shape!

smiths commented:
Should we remove the `Makefile` rule `make X_mdbook_server`?  I like that this is implemented and it is cool that we can start the mdbook server for any of the case studies from one central location, but I'm not sure the idea is consistent with the purpose of the main `Makefile`.  The main `Makefile` is for generating the examples, not for running them.  Continuing this new trend, if we have a `make X_mdbook_server`, should we also have a `make X_run` that runs the code for each example?  We have many variations of the code in some cases.  Should we have a `make` run for each variant?

Now that we have the local `Makefile` for each mdbook, I think this is more in keeping with how we've handled doing something with the examples.  The main `Makefile` generates the examples, including the "local" `Makefile`s.  The local `Makefile`s are used to run the examples, compile the pdf from tex and start the mdBook server.

I'm fine to leave `make X_mdbook_server`, but I thought we should talk about it, since if we aren't going to use this rule, we don't want to have to maintain this code going forward.

samm82 commented on code/drasil-printers/lib/Language/Drasil/HTML/Helpers.hs:
```diff
@@ -94,6 +94,10 @@ wrapGen' sepf Id s ti _ = \x ->
       te c = text $ "</" ++ c ++ ">"
   in sepf [tb s, indent x, te s] 
+wrapGen' sepf Align s ti _ = \x ->
+  let tb c = text ("<" ++ c ++ " align=\"") <> ti <> text "\">"
```
Since there is the repeated pattern of `let tb c = text ("<" ++ c ++ " <lowercase Variation>=\"")`, I'm wondering how much of these case could be generated, as well as if we could define a `show` for `Variation` to assist in this 🤔 Don't know how much effort this would be for its benefits though lol

Love this commit regardless!

BilalM04 replied:
Improved in https://github.com/JacquesCarette/Drasil/pull/3823/commits/3614102058874ba7755b208978b52b1c005356ff.

BilalM04 commented:
Removed all traces of the `Markdown GitHub` format. It has been completely moved to the [`markdownPrinter`](https://github.com/JacquesCarette/Drasil/tree/markdownPrinter) branch.

JacquesCarette commented:
Taking the plunge and merging this in!
